### PR TITLE
[Audio metrics 04/12] Add ARECHO autoregressive metric model

### PR DESCRIPTION
**Pull request by ftshijt**
Add tokenized training and autoregressive inference using published parameter names and SOS/EOS IDs. Correct the objective and reuse the repaired reference encoding behavior.

Part 4/12 of the Uni-VERSA / ARECHO revival. **Depends on [6750](https://github.com/espnet/espnet/pull/6750); keep this PR in draft until its prerequisites merge.** The series index is [#6747](https://github.com/espnet/espnet/pull/6747).

[Review only this part](https://github.com/ftshijt/espnet/compare/c456f65d3e56cd252146474e7334bc72f7b71a4e...fb7a49bdefc9611c68d064f198b972d0ced2d2e2) — 4 files and 727 changed lines added by this part.

The GitHub Files changed tab is currently cumulative because this upstream PR targets `master` and its prerequisite branches live in the fork. The focused diff above excludes preceding parts. Before marking ready, rebase onto merged prerequisites so the upstream PR contains only this component and meets the repository's small-PR limits.

Validation: Tiny ARECHO tests cover preprocessing, collation, positive finite loss, backward, and constrained categorical/numeric inference (2 tests). The complete series passes 115 local regression tests. Local execution used Python 3.10 / CPU PyTorch 2.11; supported-Python and GPU validation remain CI work. Draft status does not indicate that hosted CI has passed.

### CODE REVIEW

**Comment by ftshijt** [comment]
@coderabbitai full review

Updated to `2a5a56f191c` and merged the now-merged prerequisites, leaving four files in the PR diff. I found no posted reviews or inline comments to resolve at the start of this pass.

The direct review fixed caller-input mutation during reference normalization, missing reference-length diagnostics, target padding that ignored declared lengths, and a division-by-zero crash for empty metric targets. The entropy pass removes unused metric/type state and redundant branches while retaining checkpoint configuration compatibility and the capability flag used by the downstream CLI.

Validation: 134 component/data tests; all 16 ARECHO tests on Python 3.12; 142 full-series downstream tests; strict loading, a finite-gradient optimizer step, and numeric/categorical inference with the published ARECHO base checkpoint; pre-commit and repository-wide ESPnet2 CI lint. Eight new regression cases failed before the fixes. Hosted CI is pending for the new head.

File: espnet2/universa/ar_universa/ar_universa.py [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+
+        # NOTE(jiatong): add the metric token offset for beam search, this masking is
+        # used for pre-beam pruning
+        beam_masking = {}
+        for metric_name in metric_list:
+            metric_token = self.metric_tokenizer.get_metric_meta_label(metric_name)
+            start_idx, num_idx = self.metric_tokenizer.metric_offset[metric_name]
+            # +2 to skip the meta label token and padding token
+            start_idx = start_idx + self.metric_tokenizer.overall_offset + 2
+            end_idx = start_idx + (num_idx - 2)
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/ar_universa/ar_universa.py]: Fixed in 2703e63bb17. Both bounds now come from get_token_index(), using index 0 for numeric values and index 1 for categories, with an exclusive upper bound one past the last configured value. The duplicate metric-offset arithmetic is removed.

The new decoder regressions force an unused token to have the highest logit and verify that inference instead selects each valid numeric/categorical boundary. Both numeric cases fail on the previous code and all four pass after the fix. Validation passes 146 component/data tests, 154 downstream tests, and 28 focused cases on Python 3.12; the published checkpoint still loads strictly and passes fine-tuning/inference.

File: espnet2/universa/ar_universa/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from espnet2.universa.ar_universa.ar_universa import ARUniversa
+
+__all__ = ["ARUniversa"]
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/ar_universa/__init__.py:3]: Addressed using the documented-follow-up option. The PR description now explicitly links #6752 and states that this component alone does not provide a recipe-selectable task. #6752 contains AudioMetricTask, registers ARUniversa through ClassChoices/class_choices_list, constructs ESPnetUniversaModel, selects ARMetricProcessor/ARMetricCollateFn, and adds standard training plus legacy Universa entry points.

This separation preserves the existing small-PR series boundary rather than importing the subsequent task component here. I verified the task wiring in the follow-up and reran the full-series tests with this PR's updated files: 154 pass, including task construction and inference-CLI integration. The integration remains a required follow-up, not a capability claimed for this component PR.

**Comment by ftshijt** [comment]
@coderabbitai full review

All current inline findings and grouped suggestions are addressed in `2703e63bb17` or explicitly documented as the requested follow-up:

- **Numeric beam bounds:** derive both bounds from `get_token_index()` with the numeric/category-specific starting index. Regression tests force a higher-scoring unused token and verify both valid boundaries. The two numeric cases fail before the fix.
- **Data-parallel padding:** trim audio and reference audio to their local maximum lengths before normalization/encoding. Tests reproduce the attention shape failure before the fix with normalization both enabled and disabled, then compare padded shards with pre-trimmed inputs and verify input ownership.
- **Test organization:** moved only `test_collate_sparse_metrics_preserves_pairs` into `test_ar_universa_data.py`. Model training and empty-target integration tests remain in `test_arecho.py`, as requested in both reviews.
- **Docstring warning:** all 23 function definitions in the changed model, data, and test files now have docstrings, including the nested test factory.
- **Task integration:** the PR description explicitly links required follow-up #6752 and explains its task registration, preprocessing/collation selection, and training entry points. This component does not claim standalone recipe integration. The downstream task/CLI tests pass with the current model changes.
- **Coverage:** added meaningful tests for rejected architecture configurations and default inference setup. Both changed production modules now have 100% local statement coverage; hosted Codecov will update after CI uploads.

Validation: 146 component/data tests; 154 full-series tests; 28 focused cases on Python 3.12; strict published-checkpoint loading, a verified finite-gradient optimizer step, and numeric/categorical inference; applicable pre-commit hooks, repository-wide ESPnet2 CI lint, and diff checks all pass.

The previous documentation failure is separate from docstring coverage: [the job](https://github.com/espnet/espnet/actions/runs/35780026966/job/106923071314) built the Sphinx documentation successfully, then failed in VuePress search indexing with `SlimSearch: duplicate ID 1139`. The identical error predates this PR in the #6750 documentation logs. Fresh hosted checks are pending; I have not changed unrelated search-index infrastructure or weakened its checks.

**Comment by ftshijt** [comment]
@coderabbitai full review

Fixed the all-unlabelled batch finding from [the latest review](https://github.com/espnet/espnet/pull/6751#issuecomment-5785330356) in `b0970397629`.

Removed the conditional that omitted metric targets. The collator now always emits `metric_token` and `metric_token_lengths`; when every sample lacks `metrics`, these have shape `(B, 0)` and contain zero lengths, respectively. This uses the existing EOS-only training path without adding a second code path.

The model regression now covers both absent `metrics` keys and explicit empty dictionaries in two-sample batches, asserting empty target shapes, zero lengths, positive finite loss, zero value accuracy, and successful backward. Direct collator tests also assert the empty-target contract. Three regression cases failed before the fix.

Validation: 147 component/data tests, 155 downstream tests including the inference CLI, and 29 focused Python 3.12 cases pass. Applicable pre-commit hooks and `git diff --check` pass. Fresh hosted CI is pending.

File: espnet2/universa/ar_universa/data.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    def __call__(self, data):
+        """Collate features and variable-length metric targets independently."""
+        data = list(data)
+        keys, batch = super().__call__(
+            [
+                (uid, {k: v for k, v in sample.items() if k != "metrics"})
+                for uid, sample in data
+            ]
+        )
+        if any("metrics" in sample for _, sample in data):
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/ar_universa/data.py]: Fixed in b0970397629 by removing the conditional entirely: every batch now receives metric_token and metric_token_lengths. All-unlabelled batches produce shape (B, 0) and zero lengths. The updated regressions cover absent keys and explicit empty dictionaries in two-sample batches, finite positive EOS-only loss, zero value accuracy, backward, and the collator output contract. Three cases failed before the fix. Validation: 147 component/data tests, 155 downstream tests, and 29 focused Python 3.12 cases pass; pre-commit and diff checks pass.

**Comment by ftshijt** [comment]
Fixed the documentation CI failure in `0c45fe02639`.

The failure was a concrete route collision: the API generator emitted both `guide/espnet2/legacy/Hypothesis.md` and `guide/espnet2/legacy/_Hypothesis.md`. VuePress maps both to `/guide/espnet2/legacy/Hypothesis.html`, so SlimSearch rejected the duplicate document ID. The private tuple base was introduced with the prerequisite's shared-hypothesis fix.

The generator now excludes private classes, matching its existing exclusion of private functions. Public `Hypothesis` documentation is retained; search and CI checks stay enabled.

Validation: reproduced the collision with the installed VuePress/search packages and generated API page paths; after the fix, initialization/search indexing succeeds for 1,733 pages without duplicate routes. Two generator regressions pass, covering public/private discovery and the real beam-search classes. Lint and diff checks pass. The full hosted documentation build is running again.

@coderabbitai full review

**Comment by ftshijt** [comment]
The remaining CI failure is resolved on `0c45fe02639`.

- The hosted documentation build passes with the private-class API page fix.
- The only subsequent test failure was the existing `test_SizedDict_shared` exceeding its five-second multiprocessing timeout. All seven SizedDict tests pass locally, with the shared-process test taking 2.14 seconds.
- Reran the failed Ubuntu Python 3.12 / PyTorch 2.14 job without changing code: **57,807 passed, 907 skipped**. The Ubuntu aggregate gate passes too.
- All review threads are resolved.

Successful rerun: https://github.com/espnet/espnet/actions/runs/35838162689

**Comment by ftshijt** [comment]
Merged current master (27d0b1c4989) in 472d53cbeaf and resolved the documentation generator conflict by retaining the upstream implementation and explanation of the same private-class filtering fix. The two documentation regressions and all 29 focused ARECHO/data tests pass; diff checks pass. Fresh CI will validate the merged head.

@coderabbitai full review

**Comment by ftshijt** [comment]
Verified CI on the merged head 472d53cbeaf: all 76 checks pass, with no failed or pending checks, and the PR is mergeable.

The ASR configuration shard failed because the HuBERT large checkpoint download stopped at 93%; torch.load then reported a missing ZIP central directory. The same checkpoint had loaded successfully earlier in that job. Reran the failed shard and dependent Ubuntu gate with fresh downloads, without changing source code; both now pass.

Successful rerun: https://github.com/espnet/espnet/actions/runs/35932316444

**Review by sw005320 — APPROVED** [review]

**Comment by sw005320** [comment]
thanks a lot!
